oeps/oep-0033-bp-code-autoformatting.rst

```diff
@@ -0,0 +1,127 @@
+linters), consistent code review, or both. In general, minimizing the number
+of review comments that a reviewer has to make on an incoming change makes
+review easier, and lessens the number of back-and-forths required to merge
+the code. As such, linters that can be run locally should be preferred.
+Similarly, ambiguity in style guidelines should be minimized, so that there
+are fewer places for commentary around what amounts to personal preference
+(which might then instigate unnecessary debate).
+
+In short, more consistent, less ambiguous standards for code formatting
+make for more productive code reviews and waste less time on inconsequential
+details.
+
+Decision
+--------
+
+We will select automatic code formatters for all of our major languages. These
+formatters will have both a lint-mode (that will be run in continuous
+integration to ensure consistent formatting) and an autoformat mode (which
+can be integrated with editors to free developers from having to think about
+code formatting). These formatting standards will be applied across all
+OpenEdX-authored repositories.
+
+Selected tools should, ideally, require minimal configuration. Any required
+configuration should be added to `edx/edx-lint`_ in order to allow it to
+be centralized and standardized across repositories in the OpenEdX ecosystem.
+
+The current recommended formatters are:
+
+    - Python: `black`_
+    - Javascript: `prettier`_
+    - CSS/SASS: `prettier`_
+    - HTML: `prettier`_
+    - JSON: `prettier`_
+    - PHP: `prettier-php`_
+
+.. _black: https://github.com/ambv/black
+.. _prettier: https://prettier.io/
+.. _prettier-php: https://github.com/prettier/plugin-php
+.. _`edx/edx-lint`: https://github.com/edx/edx-lint
+
+Formatting should exposed via the standard ``Makefile`` commands
+``format`` (which should autoformat all code in the repository)
+and ``check-format`` (which should validate that all code in the
+repository is formatted). ``check-format`` should also be called
+as part of ``make quality``.
+
+When bulk-formatting the entire repo for the first time, all changes should
+be done in as few commits as possible (ideally, only one). Then, the
+commit hashes should be added to a ``.git-blame-ignore-revs`` file.
+This will allow the `git hyper-blame`_ tool to ignore them by default,
+to preserve the easy ability to see who changed particular file lines.
+
+.. _git hyper-blame: https://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/git-hyper-blame.html
+
+Consequences
+------------
+
+There will be commits to all OpenEdX repositories that cause large-scale
+reformatting without functional changes. This will muddy the blame-file
+history. Users with auto-formatters configured in their editors working
+in a repo that is not yet OEP-33 compliant will need to disable their
+autoformatter, or risk large-scale changes that are unrelated to their
+current tasks. All OpenEdX code will be formatted consistently and
```

Comment:
Do editors generally have a way to configure auto-formatting on a per-repo basis? Is that something we can configure centrally? This will help with gradual adoption of auto-formatting across Open edX. Perhaps more importantly, many developers work on non-Open-edX projects also, so universal auto-formatting is not reasonable.

Reply:
I'm not sure about all editors, but VSCode has per-workspace (aka per-repo) settings that can disable autoformatting (or, conversely, could be used to enable autoformatting only on particular repos).

Reply:
Many of my concerns can be alleviated if things were to be rolled out on a per-repo basis.